sarif/sarif.go

```diff
@@ -346,6 +360,84 @@ func (s *SarifTransformer) isSnykURI(uri string) bool {
 	return strings.HasPrefix(uri, "https_//")
 }
 
+func (s *SarifTransformer) relativePath(path string) (string, error) {
+	if s.workspacePath == "" {
+		return path, nil
+	}
+
+	if !strings.HasPrefix(path, s.workspacePath) {
+		return "", errors.Errorf(
+			"%s: result is not inside expected directory: %s",
+			path,
+			s.workspacePath,
+		)
+	}
+
+	relativePath, err := filepath.Rel(s.workspacePath, path)
+	if err != nil {
+		return "", errors.Errorf(
+			"could not get relative path from path %s using prefix %q",
+			path,
+			s.workspacePath,
+		)
+	}
+
+	return relativePath, nil
+}
+
+// normalisePath will take a given path and construct a file url pointing to
+// the file relative to the workspacePath
+func (s *SarifTransformer) normalisePath(
+	path string,
+	uriBaseId *string,
+) (*ocsf.File, error) {
+	parsedPath, err := url.Parse(path)
+	if err != nil {
+		return nil, errors.Errorf("%s: could not parse path: %w", path, err)
+	}
+
+	cleanedPath := filepath.Clean(
+		filepath.Join(parsedPath.Host, parsedPath.Path),
+	)
+
+	switch {
+	case uriBaseId == nil:
+		if s.workspacePath != "" && filepath.IsAbs(cleanedPath) {
+			relativePath, err := s.relativePath(cleanedPath)
+			if err != nil {
+				return nil, err
+			}
+
+			cleanedPath = relativePath
+		}
+	case strings.ToLower(*uriBaseId) == "%srcroot%" && filepath.IsAbs(cleanedPath):
```

Comment:
i'm not sure what %srcroot% is

Reply:
from section 3.4.4 (here)[https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/sarif-v2.1.0-errata01-os-complete.html#_Toc141790681] the way that I interpret the document is that this is a loosely defined templating variable that could be replaced with a specific directory by however is consuming the document. it is currently used by some of the tools that we integrate with. the important thing about it for me is that if it's present it signals consistently that the path we are working with is a relative one. the code expects it only as a nice to have